Add dedicated RowBlock

wenleix opened the pull request:
Ready for review.
Note the final 3 migration commits are somewhat separated arbitrarily. We can probably merge them into one given they are not too large.

    The current internal representation of a row type in Presto is an
    interleaved block wrapped by an array block. A dedicated RowBlock is
    easier to understand.

wenleix commented:
MapBlock Pull Requests (for reference)
- https://github.com/prestodb/presto/pull/7590
- https://github.com/prestodb/presto/pull/7997
- https://github.com/prestodb/presto/pull/8623

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockEncoding.java:
```diff
@@ -0,0 +1,121 @@
+        return FACTORY;
+    }
+
+    public static class MapBlockEncodingFactory
```
reminder: This name should be fixed.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java:
```diff
@@ -0,0 +1,196 @@
+    @Override
+    public String toString()
+    {
+        return format("RowBlock{SingleMapBlockWriter=%d, positionCount=%d", numColumns, getPositionCount());
```
Reminder to fix.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java:
```diff
@@ -0,0 +1,248 @@
+    @Override
+    public String toString()
+    {
+        return format("MapBlockBuilder{numColumns=%d, positionCount=%d", numColumns, getPositionCount());
```
Reminder to fix

haozhun commented:
"Remove unused method in MongoPageSource" looks good

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToRowCast.java:
```diff
@@ -126,13 +126,11 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
-                .invokeConstructor(InterleavedBlockBuilder.class, List.class, BlockBuilderStatus.class, int.class)
-                .putVariable(blockBuilder);
+        body.append(blockBuilder.set(
+                newInstance(InterleavedBlockBuilder.class,
```
first argument on its own line

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToRowCast.java:
```diff
@@ -126,13 +126,11 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
-                .putVariable(blockBuilder);
+        body.append(blockBuilder.set(
+                newInstance(InterleavedBlockBuilder.class,
+                        constantType(binder, toType).invoke("getTypeParameters", List.class),
```
`loadConstant(binder.bind(toType.getTypeParameters(), List.class))`

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java:
```diff
@@ -0,0 +1,248 @@
+    private int positionCount;
+    private int[] columnBlockOffsets;
+    private boolean[] rowIsNull;
+    private final BlockBuilder[] columnBlockBuilders;
```
Call them `field` instead of `column`

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java:
```diff
@@ -0,0 +1,248 @@
+    private BlockBuilderStatus blockBuilderStatus;
+
+    private int positionCount;
+    private int[] columnBlockOffsets;
```
Add comment that this offset is for entries and not positions.

wenleix replied:
Since it's the offset in field blocks, it kind of contain the information that the offset is for entries 

i.e. position-based offset in field block is the entry-based offset in parent block :)

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractRowBlock.java:
```diff
@@ -0,0 +1,210 @@
+public abstract class AbstractRowBlock
+        implements Block
+{
+    protected final int numColumns;
```
`numFields`

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToRowCast.java:
```diff
@@ -120,17 +120,19 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
                         constantType(binder, toType).invoke("getTypeParameters", List.class),
                         newInstance(BlockBuilderStatus.class),
-                        constantInt(toTypes.size()))));
+                        constantInt(1))));
```
What about writing `type.createBlockBuilder` instead?

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToRowCast.java:
```diff
@@ -141,23 +143,25 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
+        body.append(blockBuilder.invoke("closeEntry", BlockBuilder.class).pop());
+        body.append(constantType(binder, toType).invoke("getObject", Object.class, blockBuilder.cast(Block.class), constantInt(0))
+                .cast(Block.class)
+                .ret());
```
Format as 

```
body.append(constantType(binder, toType)
        .invoke("getObject", Object.class, blockBuilder.cast(Block.class), constantInt(0))
        .cast(Block.class)
        .ret());

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarRow.java:
```diff
@@ -31,27 +31,21 @@ public static ColumnarRow toColumnarRow(Block block)
-        Block[] fields = interleavedBlock.computeSerializableSubBlocks(interleavedBlockOffset, interleavedBlockLength);
 
-        return new ColumnarRow(block, fields);
+        return new ColumnarRow(block, fieldBlocks);
```
Looks like you intend to switch everything over at once. This is ok. I just want to check with you that this is your actual intention.

wenleix replied:
Yeah. This is the actual intention.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractRowBlock.java:
```diff
@@ -0,0 +1,210 @@
+{
+    protected final int numColumns;
+
+    protected abstract Block getColumnBlock(int columnIndex);
```
One difference is that `getBlocks` cannot be executed efficiently with `RowBlockBuilder`: it must do the element-wise cast (BlockBuilder[] -> Block[])

wenleix replied:
Never mind, array in Java is covariant! :)

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlock.java:
```diff
@@ -0,0 +1,144 @@
+        if (sizeInBytes < 0) {
+            calculateSize();
+        }
+        return sizeInBytes;
```
@haozhun : Copied from `MapBlock` :D

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractRowBlock.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    // This method should be overridden when more efficient implementation is possible.
+    protected Block[] getColumnBlocks()
```
Implementation removed. Left it as an abstract method :)

wenleix commented:
I addressed all comments except this one:

```
presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java
+    private BlockBuilderStatus blockBuilderStatus;
+
+    private int positionCount;
+    private int[] columnBlockOffsets;
      @haozhun

haozhun a day ago
Add comment that this offset is for entries and not positions.

wenleix 3 hours ago 
Since it's the offset in field blocks, it kind of contain the information that the offset is for entries
i.e. position-based offset in field block is the entry-based offset in parent block :)
We can add this as a comment ? :)
```

The reason is the position-based offset in field block is already entry-based offset in the RowBlock. But we can definitely add a comment to point this out explicitly. :)

wenleix commented:
Some related documents for the last commit (Use local variable in getSizeInBytes for Array/Map Block ):

- https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#wishful-benign-is-resilient
- https://shipilev.net/blog/2014/jmm-pragmatics/#_synchronization_order
- https://www.youtube.com/watch?v=TxqsKzxyySo&index=1&list=PLk-F_TeGx1jxX9HeIdvWkxR0t5lH-iBQL&t=572s

wenleix commented:
As discussed, we should fix getSizeInBytes by the way in https://github.com/prestodb/presto/pull/8772. As long is not atomic by JMM and 

> Programmers are encouraged to declare shared 64-bit values as volatile or synchronize their programs correctly to avoid possible complications.

(http://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.7)

fiedukow commented:
This is very good change, old version was indeed cryptic.
Thanks a lot :+1:.
